src/psbtv2.ts

```diff
@@ -0,0 +1,910 @@
+    redeemScript?:Buffer,
+    witnessScript?:Buffer,
+    bip32Derivation?:{
+      pubkey:Buffer;
+      masterFingerprint:Buffer;
+      path:string;
+    }[],
+  }) {
+    // TODO: Maybe this needs to check PSBT_GLOBAL_TX_MODIFIABLE before
+    // performing the operation.
+    const map = new Map<Key, Value>();
+    const bw = new BufferWriter();
+    bw.writeI64(amount);
+    map.set(KeyType.PSBT_OUT_AMOUNT, bw.render());
+    bw.writeBytes(script);
+    map.set(KeyType.PSBT_OUT_SCRIPT, bw.render());
+
+    if (redeemScript) {
+      bw.writeBytes(script);
+      map.set(KeyType.PSBT_OUT_REDEEM_SCRIPT, bw.render());
+    }
+
+    if (witnessScript) {
+      bw.writeBytes(script);
+      map.set(KeyType.PSBT_OUT_WITNESS_SCRIPT, bw.render());
+    }
+
+    if (bip32Derivation) {
+      for (const bip32 of bip32Derivation) {
+        bw.writeString(KeyType.PSBT_OUT_BIP32_DERIVATION, 'hex');
+        bw.writeBytes(bip32.pubkey)
+        const key = bw.render().toString('hex');
+        bw.writeBytes(bip32.masterFingerprint);
+        bw.writeBytes(parseDerivationPathNodesToBytes(bip32.path))
+        map.set(key, bw.render());
+      }
+    }
+
+    this.outputMaps.push(map);
+  }
+
+
+  // Removes an output-map from outputMaps
+  public deleteOutput (index:number) {
+    // TODO: Maybe this needs to check PSBT_GLOBAL_TX_MODIFIABLE before
+    // performing the operation.
+    const newOutputs = this.outputMaps.filter((_, i) => i !== index);
+    this.PSBT_GLOBAL_OUTPUT_COUNT = newOutputs.length;
+    this.outputMaps = newOutputs;
+  }
+
+
+  // Return the current state of the psbt as a string in the specified format.
+  public serialize (format:'base64'|'hex' = 'base64') {
+    // Build hex string from maps
+    let bw = new BufferWriter();
+    bw.writeBytes(PSBT_MAGIC_BYTES);
+    serializeMap(this.globalMap, bw);
+    
+    for (const map of this.inputMaps) {
+      serializeMap(map, bw);
+    }
+
+    for (const map of this.outputMaps) {
+      serializeMap(map, bw);
+    }
+
+    return bw.render().toString(format);
+  }
+
+
+  // Attempt to return a PsbtV2 by converting from a PsbtV0
+  static DeserializeFromV0 (psbt:string|Buffer):PsbtV2 {
+    const psbtv0Buf = bufferize(psbt);
+    const psbtv0 = Psbt.fromBuffer(psbtv0Buf);
+    const psbtv2 = new PsbtV2()
+    const psbtv0GlobalMap = psbtv0.data.globalMap
+
+    psbtv2.PSBT_GLOBAL_TX_VERSION = psbtv0.data.getTransaction().readInt32LE();
+    psbtv2.PSBT_GLOBAL_INPUT_COUNT = psbtv0.data.inputs.length;
+    psbtv2.PSBT_GLOBAL_OUTPUT_COUNT = psbtv0.data.outputs.length;
+    psbtv2.PSBT_GLOBAL_FALLBACK_LOCKTIME = 0; // Is this necessary?
+
+    for (const globalXpub of psbtv0GlobalMap.globalXpub ?? []) {
+      psbtv2.addGlobalXpub(
+        globalXpub.extendedPubkey,
+        globalXpub.masterFingerprint,
+        globalXpub.path,
+      );
+    }
+
+    let txInputs:any = [];
+    for (const [index, txInput] of psbtv0.txInputs.entries()) {
+      txInputs[index] = txInput;
+    }
```

Comment:
could we make this a const and use a map operation?

Reply:
I might be missing something obvious, but why is it that these need to be done in two steps (same for the outputs)? What's the difference between what comes back from `data.inputs` and `txInputs`? Again, probably missing something, but feels like it would be nice to do it in one step.

Separate issue and might be an over optimization, but might be nice to have these as standalone `addInputsFromPsbtV0` and `addOutputsFromPsbtV0` not a huge deal though.

Reply:
This is reading from a `bitcoinjs-lib` Psbt. Apparently not all values are available from one attribute.

Reply:
@tylerchambers, similarly, this `entries` function is returning something that is not an array prototype. It has no "map".